[stable/concourse] properly cleanup btrfs subvolume and children

cirocosta opened the pull request:
#### What this PR does / why we need it:

This PR is a follow up to https://github.com/helm/charts/pull/12398, having the `git rebase` already done.

More information about why it's needed can be found in the PR I mentioned above, but  the `tl;dr` is that a straight `rm -rf` doesn't work when what's there to be deleted is BTRFS subvolumes (rather than a plain sparse file - used when we don't have a "real" btrfs partition to use).

#### Which issue this PR fixes

  - fixes #13284

#### Special notes for your reviewer:

Please do not merge yet - I'll re-run the end-to-end tests that Concourse has first to ensure that we'll not be caught up with surprises 👍  

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

YoussB reviewed (changes requested):
@cirocosta Will you please rebase the `Chart.yml`?

smoke commented on stable/concourse/templates/worker-statefulset.yaml:
```diff
@@ -51,11 +51,16 @@ spec:
           args:
             - -ce
             - |-
+              for v in $(btrfs subvolume list --sort=-ogen {{ .Values.concourse.worker.workDir }} | awk '{print $9}'); do
```
```suggestion
              for v in $((btrfs subvolume list --sort=-ogen "{{ .Values.concourse.worker.workDir }}" || true) | awk '{print $9}'); do
```

smoke commented on stable/concourse/templates/worker-statefulset.yaml:
```diff
@@ -51,11 +51,16 @@ spec:
             - -ce
             - |-
+              for v in $(btrfs subvolume list --sort=-ogen {{ .Values.concourse.worker.workDir }} | awk '{print $9}'); do
+                btrfs subvolume delete {{ .Values.concourse.worker.workDir }}/$v
```
```suggestion
                (btrfs subvolume show "{{ .Values.concourse.worker.workDir }}/$v" && btrfs subvolume delete "{{ .Values.concourse.worker.workDir }}/$v") || true
```

smoke reviewed (changes requested):
I would suggest the following changes to gracefully continue in case the volume is not btrfs, still showing as error the fact of it.
Here is how it will workout
```
rkirilov-work-pc /home/smoke # (btrfs subvolume list /boot || true) | awk '{print $9}'
ERROR: not a btrfs filesystem: /boot
ERROR: can't access '/boot'
rkirilov-work-pc /home/smoke # echo $?
0
```

smoke commented on stable/concourse/templates/worker-statefulset.yaml:
```diff
@@ -51,11 +51,16 @@ spec:
+          securityContext:
+            privileged: true
           command:
             - /bin/sh
```
```suggestion
             - /bin/bash
```

smoke commented on stable/concourse/templates/worker-statefulset.yaml:
```diff
@@ -51,11 +51,16 @@ spec:
+              for v in $(btrfs subvolume list --sort=-ogen {{ .Values.concourse.worker.workDir }} | awk '{print $9}'); do
+                btrfs subvolume delete {{ .Values.concourse.worker.workDir }}/$v
+              done
               rm -rf {{ .Values.concourse.worker.workDir }}/*
```
```suggestion
               rm -rf "{{ .Values.concourse.worker.workDir }}/*"
```

smoke reviewed (changes requested):
Here is a working set of suggestions - same as https://github.com/concourse/charts/pull/2

Commit:
[stable/concourse] properly cleanup btrfs subvolume and children

When a `dind` (Docker in Docker) image is used with btrfs to e.g. run
integration tests as per https://hub.docker.com/r/amidos/dcind/ in some
occasions like job errors or interruptions the btrfs subvolumes are left
not cleaned.

So what happens then is that when the `rm -rf /concourse-worker-dir`
runs it fails with `Operation not permitted` error which then causes
`Init:Error` and ends in `Init:CrashLoopBackOff`.

The solution is to take that into account and properly delete all of the
btrfs subvolumes.

This can be achieved either with the suggested script or with the mount
option [user_subvol_rm_allowed](https://askubuntu.com/questions/509292/how-to-set-user-subvol-rm-allowed-capability)
that is tricky to apply or with that delete script that seems as a
better option.

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>
Co-authored-by: Radoslav Kirilov <rkirilow@gmail.com>
Co-authored-by: Taylor Silva <tsilva@pivotal.io>
